Deploy CoW DAO module on mainnet

fedgiac opened the pull request:
The contract introduced in #230 was deployed on mainnet. (Up-to-date as of commit e297f00037ed89af6458dd57c82588e897eef4d2.)

Deployment output:

<details><summary>deployment</summary>

```
$ yarn deploy --network mainnet
yarn run v1.22.22
$ hardhat deploy --network mainnet
Nothing to compile
reusing "AllocationModule" at 0x582A254713b65c140840ade25A692fBe2610682d
To enable this module in the team controller safe, execute a transaction with the following parameters on the team controller safe:
To: 0xca07EaA4253638D286caD71CBcEec11803F2709A
Value: 0
Data: 0x610b5925000000000000000000000000582a254713b65c140840ade25a692fbe2610682d
deploying "CowDaoAllocationModule" (tx: 0x68bf9d00ab29118d45644bb02a528a115b2b2bd9dc2b81079197535a159628dc)...: deployed at 0xd2D2946402c60d1C97195fa22eaD21812e1ff25D with 940371 gas
To enable this module in the CoW DAO safe, execute a transaction with the following parameters on CoW DAO:
To: 0xcA771eda0c70aA7d053aB1B25004559B918FE662
Value: 0
Data: 0x610b5925000000000000000000000000d2d2946402c60d1c97195fa22ead21812e1ff25d
updating 'networks.json'...
```
</details>

<details><summary>contract verification (already verified)</summary>

```
$ yarn verify:etherscan --network mainnet
yarn run v1.22.22
$ hardhat etherscan-verify --license LGPL-3.0 --force-license --network mainnet
already verified: AllocationModule (0x582A254713b65c140840ade25A692fBe2610682d), skipping.
already verified: CowDaoAllocationModule (0xd2D2946402c60d1C97195fa22eaD21812e1ff25D), skipping.
Done in 2.56s.
```
</details>

Relevant instructions to enable the module:

```
To enable this module in the CoW DAO safe, execute a transaction with the following parameters on CoW DAO:
To: 0xcA771eda0c70aA7d053aB1B25004559B918FE662
Value: 0
Data: 0x610b5925000000000000000000000000d2d2946402c60d1c97195fa22ead21812e1ff25d
```

### Allocation

For completeness, we generate the expected allocation transaction in the proposal.

<details><summary>Generation</summary>

Using the following CSV input ([2026-01-08-dao-vesting-revised.csv](https://github.com/user-attachments/files/24500330/2026-01-08-dao-vesting-revised.csv)):

```csv
Address,Number of Tokens,Start Date,Duration (days)
0x486c2B52FfbeD401b16ce39b7Ab9E2Ebf56D650f,"50000000",2026-02-11Z08:05:15,1460
```

and running the script:

```
$ npx hardhat start-vesting --network mainnet --allocation-module 0xd2D2946402c60d1C97195fa22eaD21812e1ff25D --csv ./2026-01-08-dao-vesting-revised.csv 
┌─────────┬──────────────────────────────────────────────┬──────────────┬────────────────────┬───────────────────┬───────────────────┐
│ (index) │ Beneficiary address                          │ Amount       │ USD cost at launch │ Vesting start     │ Vesting end       │
├─────────┼──────────────────────────────────────────────┼──────────────┼────────────────────┼───────────────────┼───────────────────┤
│ 0       │ '0x486c2B52FfbeD401b16ce39b7Ab9E2Ebf56D650f' │ '50,000,000' │ '7,500,000'        │ 'Wed Feb 11 2026' │ 'Sun Feb 10 2030' │
└─────────┴──────────────────────────────────────────────┴──────────────┴────────────────────┴───────────────────┴───────────────────┘

The allocation module needs to be enabled in the team controller Safe before new vesting positions can be added.
A transaction that enables the allocation module was added at the start of the batch.
```

we get the following transaction builder output:

```
{
  "version": "1.0",
  "chainId": "1",
  "createdAt": 1767890320560,
  "meta": {
    "name": "Start Vesting Transactions Batch",
    "description": "A transaction that batches together multiple calls to the allocation modules for starting vesting positions"
  },
  "transactions": [
    {
      "to": "0xcA771eda0c70aA7d053aB1B25004559B918FE662",
      "data": "0x610b5925000000000000000000000000d2d2946402c60d1c97195fa22ead21812e1ff25d",
      "value": "0"
    },
    {
      "to": "0xd2D2946402c60d1C97195fa22eaD21812e1ff25D",
      "data": "0x02a861a6000000000000000000000000486c2b52ffbed401b16ce39b7ab9e2ebf56d650f00000000000000000000000000000000000000000000000000000000698c383b000000000000000000000000000000000000000000000000000000000784ce00000000000000000000000000000000000000000000295be96e64066972000000",
      "value": "0"
    }
  ]
}
```
</details>

The resulting transaction builder file is the following:

[2026-01-08-cow-funding-dao-proposal.json](https://github.com/user-attachments/files/24500351/2026-01-08-cow-funding-dao-proposal.json)

A simulation can be found [here](https://dashboard.tenderly.co/public/safe/safe-apps/simulator/cb3840cc-1366-44ba-852b-e69e0a714d6c/logs).

Note that it already includes the enabling of the module discussed above, meaning that this file contains the entire proposal.

(It's expected to show a warning "This batch contains some changed properties since you saved or downloaded it" since the format used lacks the checksum.)

### Test Plan

Check that the verified contract code matches.

You can also compare the bytecode of `0xd2D2946402c60d1C97195fa22eaD21812e1ff25D` and `0x582A254713b65c140840ade25A692fBe2610682d` to confirm that there are no changes outside of the deployment parameters.

<details><summary>bytecode diff</summary>

<img width="1592" height="803" alt="image" src="https://github.com/user-attachments/assets/d58e4aad-1ad1-44fd-afd1-38edb471b8c8" />

</details>

fedgiac committed:
Update hardhat-deploy

fedgiac committed:
Add module for CoW DAO

fedgiac committed:
Update description

fedgiac committed:
Deploy CoW DAO module on mainnet

fedgiac committed:
Separate deployment names from contract name (and delete unused contract)

fedgiac committed:
Merge branch 'cow-dao-module' into deployment-cow-dao-module

fedgiac committed:
Remove deployment

fedgiac committed:
Deploy CoW DAO module on mainnet

kaze-cow commented on deployments/mainnet/solcInputs/a646833678744341c349f0a109c850d1.json:
worth mentioning that this new `solcInputs` file is almost the same as the original (already checked into the `main` branch) save for one difference

```
--- deployments/mainnet/solcInputs/a646833678744341c349f0a109c850d1.json	2026-01-13 19:15:32.376531854 +0900
+++ deployments/mainnet/solcInputs/bcd8532b3dae86997a3aae16f3b66a58.json	2026-01-13 19:04:32.597336281 +0900
@@ -7,9 +7,6 @@
     "src/contracts/interface/CowProtocolTokens.sol": {
       "content": "// SPDX-License-Identifier: LGPL-3.0-or-later\npragma solidity ^0.8;\n\n/// @dev Interface exposing some of the functions of the governance token for the CoW Protocol.\n/// @title CoW Protocol Governance Token Minimal Interface\n/// @author CoW Protocol Developers\ninterface CowProtocolToken {\n    /// @dev Moves `amount` tokens from the caller's account to `to`.\n    /// Returns true. Reverts if the operation didn't succeed.\n    function transfer(address to, uint256 amount) external returns (bool);\n}\n\n/// @dev Interface exposing some of the functions of the virtual token for the CoW Protocol.\n/// @title CoW Protocol Virtual Token Minimal Interface\n/// @author CoW Protocol Developers\ninterface CowProtocolVirtualToken {\n    /// @dev Converts an amount of (virtual) tokens from this contract to real\n    /// tokens based on the claims previously performed by the caller.\n    /// @param amount How many virtual tokens to convert into real tokens.\n    function swap(uint256 amount) external;\n\n    /// @dev Address of the real COW token. Tokens claimed by this contract can\n    /// be converted to this token if this contract stores some balance of it.\n    function cowToken() external view returns (address);\n}\n"
     },
-    "src/contracts/test/MockVCow.sol": {
-      "content": "// SPDX-License-Identifier: LGPL-3.0-or-later\npragma solidity ^0.8;\n\n/// @dev Mock of vCOW token. Emits an event when calling `swap`.\n/// @title vCOW mock contract\n/// @author CoW Protocol Developers\ncontract MockVcow {\n    address public cowToken;\n\n    constructor(address _cowToken) {\n        cowToken = _cowToken;\n    }\n\n    event Swapped(address caller, uint256 amount);\n\n    function swap(uint256 amount) external {\n        emit Swapped(msg.sender, amount);\n    }\n}\n"
-    },
     "src/contracts/vendored/Enum.sol": {
       "content": "// SPDX-License-Identifier: LGPL-3.0-only\n\n// Vendored from @gnosis.pm/safe-contracts v1.3.0, see:\n// <https://raw.githubusercontent.com/gnosis/safe-contracts/v1.3.0/contracts/common/Enum.sol>\n\npragma solidity >=0.7.0 <0.9.0;\n\n/// @title Enum - Collection of enums\n/// @author Richard Meissner - <richard@gnosis.pm>\ncontract Enum {\n    enum Operation {\n        Call,\n        DelegateCall\n    }\n}\n"
     },
```

I think functionally its the same as what ends up on mainnet down to the CBOR, but wouldn't it be nice to reuse the existing artifact?

fedgiac replied:
Interesting, thanks for pointing this out, I didn't see it.
I agree it should work without this file, but let's keep it since it was automatically generated by the deployment script.

kaze-cow reviewed (approved):
inclusion of the `solcInputs` file was a bit surprising to me as the inputs should be the same, but turns out there is one small difference in the inclusion of the `MockVCow` test file in the compilation output. I don't think this harms anything, but interesting to see.

deployment itself seems good. initcode is the same as the original deployment except for the Safe address, and addresses match up.

anxolin reviewed (approved):
Everytying worked as described in the instructions. Thanks for the detailed steps and setting expectations!

ubernit: not from this PR, but I wish `start-vesting` would have written the output file nam in the console log. I check in the script cause at first I didn't know where to look

fedgiac committed:
Clarify usage

fedgiac committed:
Mention constants in readme

fedgiac committed:
COW_DAO → COW_DAO_SAFE

fedgiac committed:
Move deployed code to dedicated function

fedgiac committed:
Move deployment function to deployment files to work around a typing issue

CI was reverting while the function was located in the library files.
Moving it to the deployment files is a hack that fixes the issue.
I'm not really sure what's going on, but it looks like the type I need to use comes from type extensions and `declare module` so I'm not sure how to access them directly.

Relevant issue link:
https://github.com/wighawag/hardhat-deploy-ethers/issues/18

fedgiac committed:
Fix missing variable deployment name

fedgiac committed:
Merge remote-tracking branch 'origin/main' into cow-dao-module

fedgiac committed:
Merge branch 'cow-dao-module' into deployment-cow-dao-module
